miniml/compiler/compile.scm

```diff
@@ -1725,6 +1767,37 @@
          (nenv-modules (add-bindings (env-get-modules env) (env-get-modules menv))))
     (mkenv nenv-vars nenv-constrs nenv-fields nenv-modules)))
 
+(define (mkstruct env defs)
+  (let* ((mark (lambda (e) (vhash-map (lambda (k v) (cons #f (cdr v))) e)))
+         (modenv (mkenv (mark (env-get-vars env))
+                        (mark (env-get-constrs env))
+                        (mark (env-get-fields env))
+                        (mark (env-get-modules env)))))
+    (list 'VModule (compile-defs modenv defs))))
+
+(define (apply-functor mod arg)
+  (assert (equal? (car mod) 'VFunctor))
+  (let* ((fct (car (cdr mod)))
```

Comment:
If you use pattern-matching instead of an assertion+cadr here,

```
(match-let* ((('VFunctor fct) mod)
             (nenv ...))
```

you get the following side-benefits;
- (minor) a trace in the code of the expected number of arguments of the `'Vfunctor` constructor
- (major) if an input does not fit here, you get a proper match-failure error from guile that shows the offending input, instead of a less-informative assertion failure; this has really helped me debug compiler code